### DISCUSSION TITLE
Laser ttl 20khz

### DISCUSSION THREAD

**Discussion by djidai** [arkypita/LaserGRBL#890] [Q&A]
Hi every one 
I have  driver which has parameters ttl 20 khz. When i try engraving wit  gray tone. I have only 4 colors black, white, and 2 grade of gray. I change frequency in grbl from 1 to 8 khz but i didn't have any changes. Wisely i se what laser change power, but i don't  have  grade in result. Any idea?

**Comment by djidai** [comment]
Ohhh i forgot attach the foto of my driver.
![DDAE32D2-899D-4C0F-A7AF-57318E001F47](https://user-images.githubusercontent.com/61481360/75298838-7354d980-582b-11ea-9c3c-1960339811b5.jpeg)
 I bout it hear 
https://www.ebay.com/itm/5A-SCU-Blue-Laser-Driver-12VDC-TTL-For-NDB7A75-NUBM44-NUBM05-NUBM-06-07-08-46-49-/263801328863

**Comment by StuartB4** [comment]
I have the same driver on my laser. When I bought it it said on the website that the PWM signal is inverted on this driver.
![IMG_0784](https://user-images.githubusercontent.com/16980026/75302411-9dab9480-5835-11ea-9ee0-005956b38627.JPG)


Which means the laser is On at full power when the PWM signal is 0 volts and Off when it's 5 volts.
That is completely opposite to most lasers which are Off with 0 volts and On at full power with 5 volts.
Yours may be the same because I just saw this on the link to the ebay listing where you bought it.

> 5.Intigrate TTL controal slot/0V on and 5V 20kHZ.


If it is you need to invert the PWM signal. 
There are a couple of ways of doing it.
One is to use a transistor/resistor circuit like this one:
https://blog.eccentricworkshop.com/grbl-pwm-inverter-circuit

The one that worked for me was to edit the config.h file in GRBL, then compile it and reupload it to the Arduino.

**Comment by djidai** [comment]
but I don’t have any problems, it’s turned on if I don’t do anything and turns on when I give this task

**Comment by StuartB4** [comment]
Oh ok, must be 0v =Off and 5v On then like most lasers.
Do you have a picture of your grbl configuration settings?.

**Comment by djidai** [comment]
Yes.
![20200227_084308](https://user-images.githubusercontent.com/61481360/75431164-5829bd00-5944-11ea-8a1c-1b3538d19d0c.jpg)
![20200227_084318](https://user-images.githubusercontent.com/61481360/75431175-5a8c1700-5944-11ea-8221-7d2b00b4ab90.jpg)
And I did some changes in files, turned off the z axis and changed the PWM modulation frequency

**Comment by djidai** [comment]
and yes I controlled the laser through both m3 and m4, but the result is as I wrote earlier.

**Comment by StuartB4** [comment]
Do you have the S-Min and S-Max setting the same in Target Image settings as it is in Grbl $30 and $31.

![U7878787ntitled](https://user-images.githubusercontent.com/16980026/75453155-5fb08c80-596b-11ea-9d0c-b2892b3d9aa1.jpg)

![565656565](https://user-images.githubusercontent.com/16980026/75453175-67703100-596b-11ea-8642-2971f680703d.jpg)

**Comment by StuartB4** [comment]
I noticed you wrote this earlier. 

> but I don’t have any problems, it’s turned on if I don’t do anything and turns on when I give this task

Did you mean it's OFF if you don't do anything?

**Comment by djidai** [comment]
> I noticed you wrote this earlier.
> 
> > but I don’t have any problems, it’s turned on if I don’t do anything and turns on when I give this task
> 
> Did you mean it's OFF if you don't do anything?


Yes.

**Comment by djidai** [comment]
> Do you have the S-Min and S-Max setting the same in Target Image settings as it is in Grbl $30 and $31.
> 
> ![U7878787ntitled](https://user-images.githubusercontent.com/16980026/75453155-5fb08c80-596b-11ea-9d0c-b2892b3d9aa1.jpg)
> 
> ![565656565](https://user-images.githubusercontent.com/16980026/75453175-67703100-596b-11ea-8642-2971f680703d.jpg)

Yes, I changed these parameters in different ways, but still I do not get an adequate result.

**Comment by djidai** [comment]
how did you invert pwm? in a soft?

**Comment by StuartB4** [comment]
You have to change it in GRBL before you flash it on to the chip on the controller board.
You can't just change it in software because it gets stored in the chip.
I just changed one line in config.h file, then had to compile grbl and upload it to the chip.

You have to remove the 2 forward slashes // from the start of the line, then save the file, compile it and send it to the chip.
This is what it was before I changed it:

![Unererertitled](https://user-images.githubusercontent.com/16980026/75591150-e900b580-5a76-11ea-912e-6bf016d3b4a7.jpg)

 and after it looks like this:

![Unererertitled2](https://user-images.githubusercontent.com/16980026/75591184-f7e76800-5a76-11ea-9a93-a1aae9d9d037.jpg)

**Comment by djidai** [comment]
I tried it, but nothing has changed. I have a default variable spindle.
Sorry you have a DSO Shild or oscilloscope.
you can see the parameters, signal and voltage values ​​after the driver, i.e. what goes on ld?
maybe my driver is faulty.
also what ttl frequency do you use?
![20200229_003927](https://user-images.githubusercontent.com/61481360/75597214-1a37b080-5a8c-11ea-9d5b-7bc4b38be8b5.jpg)
here are my tests for today with different parameters and more.
nothing has changed and at best I have only 4 colors, black, white and about 2 steps of grad.

**Comment by djidai** [comment]
![20200229_004440](https://user-images.githubusercontent.com/61481360/75597317-c6799700-5a8c-11ea-8582-642698892208.jpg)
![20200229_004421](https://user-images.githubusercontent.com/61481360/75597325-c9748780-5a8c-11ea-8ff5-e8826ca09d29.jpg)
This is my board and driver.
the driver is connected to the gnd and d11 board. It receives power from an external 12 volt 12.5 ampere power supply that is not connected to the system.

**Comment by StuartB4** [comment]
> Sorry you have a DSO Shild or oscilloscope.

No sorry I don't. I have a digital multimeter but it's faulty and does'nt give proper readings, the digits change to fast to see them.

I use 8kHz frequency, it's actually only 7.8 really.

![Unererertitled3](https://user-images.githubusercontent.com/16980026/75598974-6852b100-5a98-11ea-8288-c0959931311d.jpg)

So you have 2 power supplies ?
If you use 2 you should connect the ground on the output leads of both of them together.
I use one power supply for everything, it came from an old PC, loads of amps.

Looking at the test piece you posted it does'nt look as bad as I thought it would.
I've done tests myself with similar results, but when engraving an image with all the settings as good as I could get them it turned out ok. A good test is to engrave a photo on something black like slate. Invert the image first so it looks like a negative.  These are on slate with different settings for each one. When you look close they don't actually have loads of different grey scales.

![IMG_0818](https://user-images.githubusercontent.com/16980026/75599461-f4ff6e00-5a9c-11ea-87d9-0a2f02fd4439.JPG)
![IMG_0819](https://user-images.githubusercontent.com/16980026/75599462-f6c93180-5a9c-11ea-88e4-82ffa78e155d.JPG)
![IMG_0820](https://user-images.githubusercontent.com/16980026/75599465-f892f500-5a9c-11ea-822f-151881e380a8.JPG)
![IMG_0821](https://user-images.githubusercontent.com/16980026/75599467-faf54f00-5a9c-11ea-9f36-dd83c1a05822.JPG)

My system uses a Protoneer CNC shield and a genuine Arduino UNO with GRBL 1.1h, and DRV8825 stepper drivers. I have 2 limit switches one on X and one on Y axis, just used for Homing.

**Comment by StuartB4** [comment]
I just noticed you have the 3 pin connector on the board in the picture you posted.
That is usually to connect a 3 wire laser.
+pin=12v,  -pin= ground and S=PWM signal.
Do you have a mutltimeter so you can you check to see if the S pin is connected to pin 11 on the Nano board. If it is then you could use that.

![Unuuuuuutitled](https://user-images.githubusercontent.com/16980026/75606106-75e05900-5ae1-11ea-9214-855a3651fb89.jpg)

**Comment by djidai** [comment]
> I just noticed you have the 3 pin connector on the board in the picture you posted.
> That is usually to connect a 3 wire laser.
> +pin=12v, -pin= ground and S=PWM signal.
> Do you have a mutltimeter so you can you check to see if the S pin is connected to pin 11 on the Nano board. If it is then you could use that.
> 
> ![Unuuuuuutitled](https://user-images.githubusercontent.com/16980026/75606106-75e05900-5ae1-11ea-9214-855a3651fb89.jpg)

my laser is connected.  but I take s and gnd from this connector.
+ - 12 volts I take from an external power supply.
![20200229_111304](https://user-images.githubusercontent.com/61481360/75606427-85ad6c80-5ae4-11ea-96c0-6fa93286a8f7.jpg)
![20200229_111225](https://user-images.githubusercontent.com/61481360/75606430-880fc680-5ae4-11ea-99e1-ee43d53de889.jpg)
![20200229_111210](https://user-images.githubusercontent.com/61481360/75606431-8940f380-5ae4-11ea-9062-d39dc72fae30.jpg)
I tried to unify gnd by setting a jumper as in the photo, but it did not give any changes.
also my board has a pwm plug but there is a 12 volt siren wire and i don't use it.
about the multimeter and pwm.  your multimeter is not broken, it just does not understand pwm.  I will use such a cheap device for this.

![20200229_111925](https://user-images.githubusercontent.com/61481360/75606533-9f02e880-5ae5-11ea-9a68-0e61eee91e43.jpg)
You can buy it on togging platforms even a kit for self-assembly.

**Comment by djidai** [comment]
I contacted the seller of the driver and here is his answeres. it work like 0V on and 5V off.but in that case, I don’t understand how it works for me.
it works for me 0 off 5 on.
nonsense

**Comment by StuartB4** [comment]
Oh right, that is strange. Mine was definately 0 on and 5v off. 
Until I changed the setting in config.h it was on and stayed on as soon as I switched the machine on.
Mine was already connected to a laser because I bought the whole thing together from Wish.com.
It was supposed to be 5 watts but when it arrived it was 7. I also got it half price because I had a discount code.

![IMG_0826](https://user-images.githubusercontent.com/16980026/75608956-396e2680-5afc-11ea-9adf-9099f6d673cc.JPG)

![IMG_0827](https://user-images.githubusercontent.com/16980026/75608957-3a9f5380-5afc-11ea-8ca3-184e57985a50.JPG)

![IMG_0828](https://user-images.githubusercontent.com/16980026/75608958-3d9a4400-5afc-11ea-920e-b0fd59eada76.JPG)

![66666](https://user-images.githubusercontent.com/16980026/75608890-8f8e9a00-5afb-11ea-9157-aa92f0095813.jpg)

![77777](https://user-images.githubusercontent.com/16980026/75608892-93bab780-5afb-11ea-976d-5f84b97cf672.jpg)

My multimeter is definately faulty, it does random characters on the display when I test AC or DC Voltage and Current. The only thing it's ok with is resistance, diode test and continuity beeper.
Need a new one, had it about 25 years and it's never been calibrated since I got it.

**Comment by djidai** [comment]
my friend helped me invert the signal where 0 is 100 and 5 is off.
it does not turn on in idle mode. but now I have a different problem, laser grbl creates the code for normal mode where 0 is off and 5 is on.
but anyway I tried engraving gradient, and it was not successful.
I have no idea what to do with it now.

**Comment by StuartB4** [comment]
>  my friend helped me invert the signal where 0 is 100 and 5 is off.

How did your friend invert it?. 
If it's inverted it should be completely opposite, not 100.
It should be 0 and 5, Off and Full on. 

If the people you bought it from say it should be 0=On and 5v=Off then it must be right.
My driver is the same driver so I don't see why they would be different.

**Comment by djidai** [comment]
he edited the file cpu_map

**Comment by StuartB4** [comment]
Oh right, ok.

**Comment by djidai** [comment]
I replaced the driver with the exact one that was in stock.  but contrary to my hopes, nothing has changed.
you have access to your driver, I'm interested in a photograph of the back of the board.

**Comment by StuartB4** [comment]
I also asked the seller on ebay how it works and got a completely different answer. I don't think they know how it works. Here's the reply I got. You can see my question at the bottom.

![rtrtr](https://user-images.githubusercontent.com/16980026/75627189-0c864600-5bc6-11ea-8412-3ad1ef951aa4.jpg)

**Comment by djidai** [comment]
apparently it all depends on the weather on Mars.
it blows my brain!!!

**Comment by StuartB4** [comment]
Yeah I know what ya mean. I'll get a picture of the undeside of mine in a minute.

**Comment by StuartB4** [comment]
![IMG_0837](https://user-images.githubusercontent.com/16980026/75630111-c6d67700-5bdf-11ea-960c-39c3611c43d3.JPG)

**Comment by djidai** [comment]
Thanks for the photo.  it's the same as on my board.

**Comment by djidai** [comment]
![20200303_173730](https://user-images.githubusercontent.com/61481360/75812889-e8259780-5d86-11ea-933a-a792186fc943.jpg)

I conducted a number of tests, with different speeds and settings.

in fact, I have only 2 colors, these are black and white.
I do not understand what is wrong.

**Comment by djidai** [comment]
I foun it on ebay.
https://rover.ebay.com/rover/0/0/0?mpre=https%3A%2F%2Fwww.ebay.com%2Fulk%2Fitm%2F232923988707
The main interest in this is the description that contains information about the frequency of the supplied signal ttl.
so I’m pretty sure that you need to expand the spectrum of modulated frequencies in grbl.
I spent a series of dough with my driver, grbl and dso.After the signal is processed by the driver, its shape changes, as well as other parameters, the signal becomes weaker.  but in version 1 and 2 kiloherkha it is vaguely reminiscent of a square, in version 7 kHz it has a terrible shape and looks like ^ this.
Any ideas how you can increase the frequency to 20 kilohertz?

or even more?

**Comment by djidai** [comment]
I found datashit for the main cheap on the driver.
https://www.rlocman.ru/datasheet/data.html?di=167804&/TPS92641PWP
And look on this 
![Screenshot_20200308-032424_Chrome](https://user-images.githubusercontent.com/61481360/76166088-bf6d1b80-6153-11ea-9106-bf8bbd0e5fcf.jpg)

**Comment by HGadotti** [comment]
Dear friend, did you get the solution for your problem? I would like to buy this driver to assemble a homemade laser. but now I am afraid to use this driver.

**Comment by mundsen** [comment]
I ordered one (analog = PWM)

There is 2 different type, one TTL and one analog/PWM

Analog: Intigrate Analogue controal slot----0V to 5V ---weak to strong.Frequncy is 20kHZ.

TTL: Intigrate TTL controal slot/0V on and 5V 20kHZ.

I use a controller with PWM out, so I ordered the analog - hope that is going to work
